### PR TITLE
Completed Advanced Training

### PR DESCRIPTION
**Pull request by navjotbrar**

### CODE REVIEW

File: Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp [hunk 1]
```diff
@@ -8,6 +8,9 @@ namespace
 BatteryStateOfChargeService::BatteryStateOfChargeService(double initialStateOfChargePercent)
 : initialStateOfChargePercent_(initialStateOfChargePercent)
 {
+    ampHourUsed=0.0;
+    holdC = QTime(0,0,0,0);
+    holdD = QTime(0,0,0,0);
```
Comment (Reddragoneer1) [inline Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp]: What is holdC and holdD, I think it stands for charging vs depleting but can you make it so. For example, isCharging and isDepleting?

File: Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp [hunk 2]
```diff
@@ -16,23 +19,53 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
...
+        isCharging_=true;
+    }else{
+        isCharging_=false;
+    }
+
+    if(ampHourUsed==0.0){
+        ampHourUsed=BATTERY_AMP_HOUR_CAPACITY*(initialStateOfChargePercent_/100);
+    }
+    else{
+        ampHourUsed+=((prevCurr+batteryData.current)/2)*((prevTime.msecsTo(batteryData.time))/3600000.0) ;
```
Comment (Reddragoneer1) [inline Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp]: Change 3600000.0 to constant

Comment (bill-luu) [reply Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp]: For further elaboration, the value `3600000.0` is something more of a "Magic Number". Just looking at the value, it's not immediately obvious that it represents the number of seconds in an hour. 

Typical convention is to put it in a constant value in a namespace, all caps: 
`const double MY_CONSTANT_VALUE = 3600000.0"

File: Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp [hunk 3]
```diff
@@ -30,9 +30,9 @@ bool BatteryStateOfChargeService::isCharging() const
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
     if(isCharging_){
-        return holdC;
+        return timeUntilCharged;
     }else{
```
Comment (chordatgh1234) [inline Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp]: Try to keep open and closed curly brackets on separate lines for readability. 
like so: 
`return timeUntilCharged;`
`}`
`else { `

Comment (chordatgh1234) [inline Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp]: Try to keep open and closed curly brackets on separate lines for readability.
like so: 
`return timeUntilCharged;`
`}`
`else { `

File: Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp [hunk 4]
```diff
@@ -43,17 +43,19 @@ void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
     // This is where you can update your variables
     // Hint: There are many different ways that the totalAmpHoursUsed can be updated
     // i.e: Taking a running average of your data values, using most recent data points, etc.
+    const double ms_hour_conversion = 3600000.0;
```
Comment (chordatgh1234) [inline Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp]: Common convention is to declare magic constant numbers in the namespace section at the top of the file. This way, all functions within the file have access to the constant if needed.
Make sure you use all uppercase for the constant numbers.

File: Recruit-Training/Advanced-Recruit-Training/src/BatteryStateDisplayService.cpp [hunk 5]
```diff
@@ -27,7 +32,9 @@ void BatteryStateDisplayService::handleBatteryDataReceived(const BatteryData& ba
 
     QTextStream(stdout) << "Voltage: " << batteryData.voltage
         << " Current: " << batteryData.current
-        << " Total Ah used: " << batteryStateOfChargeService_.totalAmpHoursUsed() << endl;
+        << " Total Ah used: " << batteryStateOfChargeService_.totalAmpHoursUsed()
+        << " Time Until" << (batteryStateOfChargeService_.isCharging()? " Charged: " : " Depleted: ")
+        << batteryStateOfChargeService_.timeWhenChargedOrDepleted().toString(STRING_TIME_FORMAT) << endl;
 
     // TODO: Print out time until the battery is fully charged or depleted.
```
Comment (Reddragoneer1) [inline Recruit-Training/Advanced-Recruit-Training/src/BatteryStateDisplayService.cpp]: For style, you can delete the comments, especially TODO ones if they've been done.

File: Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp [hunk 6]
```diff
@@ -1,13 +1,18 @@
...
 {
     const double BATTERY_AMP_HOUR_CAPACITY = 123.0;
+    const double MS_HOUR_CONVERSION = 3600000.0;
+    const int PERCENT_TO_DECIMAL = 100;
 }
 
 BatteryStateOfChargeService::BatteryStateOfChargeService(double initialStateOfChargePercent)
 : initialStateOfChargePercent_(initialStateOfChargePercent)
 {
+    ampHourUsed=0.0;
```
Comment (bill-luu) [inline Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp]: Minor thing, put spaces between operators.
`ampHourUsed = 0.0`

File: Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp [hunk 7]
```diff
@@ -1,13 +1,18 @@
...
+#include "BatteryData.h"
 namespace
 {
     const double BATTERY_AMP_HOUR_CAPACITY = 123.0;
+    const double MS_HOUR_CONVERSION = 3600000.0;
+    const int PERCENT_TO_DECIMAL = 100;
 }
 
 BatteryStateOfChargeService::BatteryStateOfChargeService(double initialStateOfChargePercent)
 : initialStateOfChargePercent_(initialStateOfChargePercent)
```
Comment (bill-luu) [inline Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp]: Use an [initializor list](https://en.cppreference.com/w/cpp/language/initializer_list) for private variables in a constructor.

File: Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp [hunk 8]
```diff
@@ -16,23 +21,50 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
...
+    }
 }
 
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
 {
-    Q_UNUSED(batteryData);
-    // This is where you can update your variables
-    // Hint: There are many different ways that the totalAmpHoursUsed can be updated
-    // i.e: Taking a running average of your data values, using most recent data points, etc.
+    if(batteryData.current<0){
```
Comment (bill-luu) [inline Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp]: Team convention is to start braces on a new line

File: Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.h [hunk 9]
```diff
@@ -25,4 +25,12 @@ class BatteryStateOfChargeService : public I_BatteryStateOfChargeService
     const double initialStateOfChargePercent_;
 
     // Add your necessary variables here
+    bool isCharging_;
+    double prevCurr;
```
Comment (bill-luu) [inline Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.h]: Team convention is to leave a trailing underscore at the end of a private variable.

File: Recruit-Training/Advanced-Recruit-Training/src/LogFileReader.cpp [hunk 10]
```diff
@@ -51,21 +52,31 @@ bool LogFileReader::readAll(const QString& fileName)
...
+    if(!ok){
+        return false;
+    }
+    batteryData.current = sections.at(2).toDouble(&ok);
+    if(!ok){
+        return false;
+    }
 
-    return true;
+    if(batteryData.time.isValid()){
```
Comment (bill-luu) [inline Recruit-Training/Advanced-Recruit-Training/src/LogFileReader.cpp]: The above 3 if statements can be combined into one.

File: Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp [hunk 11]
```diff
@@ -16,23 +21,50 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
...
+    if(hourCharge<0){
+        hourCharge = hourCharge*-1;
+    }
+    hourDep = (BATTERY_AMP_HOUR_CAPACITY-ampHourUsed)/batteryData.current;
+    QTime hourUntilCharged = QTime(0,0,0,0);
+    QTime hourUntilDepleted = QTime(0,0,0,0);
+    hourUntilCharged = hourUntilCharged.addMSecs((int)(hourCharge*MS_HOUR_CONVERSION));
+    hourUntilDepleted = hourUntilDepleted.addMSecs((int)(hourDep*MS_HOUR_CONVERSION));
+    timeUntilCharged = hourUntilCharged;
+    timeUntilDepleted = hourUntilDepleted;
```
Comment (bill-luu) [inline Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp]: Lines 51 through 69 look very busy. It's helpful to add in new lines to separate some ideas such as variable creation, calculations, etc , for example, I would recommend something like this:
```
    if(ampHourUsed==0.0)
    {
        ampHourUsed=BATTERY_AMP_HOUR_CAPACITY*(initialStateOfChargePercent_/PERCENT_TO_DECIMAL);
    }
    else
    {
        ampHourUsed+=((prevCurr+batteryData.current)/2)*((prevTime.msecsTo(batteryData.time))/MS_HOUR_CONVERSION) ;
    }

    prevCurr = batteryData.current;
    prevTime = batteryData.time;

    hourCharge = ampHourUsed/batteryData.current;

    if(hourCharge<0)
    {
        hourCharge = hourCharge*-1;
    }

    hourDep = (BATTERY_AMP_HOUR_CAPACITY-ampHourUsed)/batteryData.current;

    QTime hourUntilCharged = QTime(0,0,0,0);
    QTime hourUntilDepleted = QTime(0,0,0,0);

    hourUntilCharged = hourUntilCharged.addMSecs((int)(hourCharge*MS_HOUR_CONVERSION));
    hourUntilDepleted = hourUntilDepleted.addMSecs((int)(hourDep*MS_HOUR_CONVERSION));

    timeUntilCharged = hourUntilCharged;
    timeUntilDepleted = hourUntilDepleted;
```

This bunching helps a lot with readability, and will let the reader process the code in their head a little bit easier :)

File: Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp [hunk 12]
```diff
@@ -16,23 +21,50 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
...
-    // i.e: Taking a running average of your data values, using most recent data points, etc.
+    if(batteryData.current<0){
+        isCharging_=true;
+    }
+    else{
+        isCharging_=false;
+    }
+
+    if(ampHourUsed==0.0){
+        ampHourUsed=BATTERY_AMP_HOUR_CAPACITY*(initialStateOfChargePercent_/PERCENT_TO_DECIMAL);
```
Comment (bill-luu) [inline Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp]: This should go in the constructor.

File: Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp [hunk 13]
```diff
@@ -16,23 +21,50 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
...
+    }
+    else{
+        ampHourUsed+=((prevCurr+batteryData.current)/2)*((prevTime.msecsTo(batteryData.time))/MS_HOUR_CONVERSION) ;
```
Comment (bill-luu) [inline Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp]: This single line does a lot, I would recommend splitting complex calculation it into multiple ones for readability.

**Pull request closed without merging**